pop_tools/zonal_average.py

```diff
@@ -0,0 +1,247 @@
+
+        dx: float
+          Horizontal grid spacing in x-direction for output grid in degrees
+
+        dy: float
+          Horizontal grid spacing in y-direction for output grid in degrees
+
+        mask: `xarray.Dataarray`
+          User defined region mask
+
+        regrid_method: string
+          Regridding method to be used within xESMF (default is conservative)
+
+        method_gen_grid: string
+          Method used to generate the output grid - default is a regular lat/lon grid
+        """
+        if grid_name is not None:
+            self.grid_name = grid_name
+
+            # Use pop-tools to retrieve the grid
+            self.grid = get_grid(grid_name)
+
+        elif grid is not None:
+            self.grid = grid
+
+        else:
+            raise ValueError('Failed to input grid name or grid dataset')
+
+        # Set the dx/dy parameters for generating the grid
+        self.dx = dx
+        self.dy = dy
+
+        # Set the regridding method
+        self.regrid_method = regrid_method
+
+        # Set the grid generation method
+        self.method_gen_grid = method_gen_grid
+
+        # If the user does not input a mask, use default mask
+        if not mask:
+            self.mask = self.grid['REGION_MASK']
+
+            if 'region_name' not in self.grid.variables:
+                self.mask_labels = np.unique(self.mask.values)
+
+            else:
+                self.mask_labels = self.grid['region_name']
+
+        else:
+            self.mask = mask
+
+    # Setup method for regridding a dataarray
+    def _regrid_dataarray(self, da_in, regrid_mask=False, regrid_method=None):
+
+        src_grid = _convert_to_xesmf(da_in, self.grid)
+        dst_grid = _generate_dest_grid(self.dy, self.dx, self.method_gen_grid)
+
+        # If the user does not specify a regridding method, use default conservative
+        if regrid_method is None:
+            regridder = _generate_weights(src_grid, dst_grid, self.regrid_method)
+
+        else:
+            regridder = _generate_weights(src_grid, dst_grid, regrid_method)
+
+        # Regrid the input data array, assigning the original attributes
+        da_out = regridder(src_grid[da_in.name])
+        da_out.attrs = da_in.attrs
+
+        return da_out
+
+    def regrid(self, obj, **kwargs):
+        """generic interface for regridding DataArray or Dataset"""
+        if isinstance(obj, xr.Dataset):
+            ds_list = []
+            for var in obj.variables:
+
+                # Make sure the variable has the correct dimensions, is not a coordinate, and is not a velocity
+                if (
+                    ('nlat' and 'nlon' in obj[var].dims)
+                    and (var not in list(obj.coords))
+                    and ('velocity' not in obj[var].long_name.lower())
+                ):
+                    ds_list.append(obj[var])
+            return xr.merge(ds_list).map(self._regrid_dataarray, keep_attrs=True, **kwargs)
```

Comment:
instead build a list of variables names and then `ds[list_of_varnames].map(...)`